src/v2/features/characters/characters-value.service.ts

```diff
@@ -0,0 +1,488 @@
+import React, { useState, useMemo } from 'react';
+import { ICharacter2, IMaterialEstimated2, IMaterialFull, IMaterialRecipeIngredientFull } from 'src/models/interfaces';
+import { orderBy, sortBy, sum } from 'lodash';
+import { Rank, Rarity, RarityStars } from 'src/models/enums';
+import rankUpData from 'src/assets/rankUpData.json';
+import recipeData from 'src/assets/recipeData.json';
+import { StaticDataService } from 'src/services/static-data.service';
+import { getEnumValues } from 'src/shared-logic/functions';
+
+export class CharactersValueService {
+    public static getCharacterValue(character: ICharacter2): number {
+        if (character.rank === Rank.Locked) {
+            return 0;
+        }
+
+        const valueLevel =
+            // CharactersValueService.getUnlockValue(character.initialRarity, character.name) +
+            // CharactersValueService.getExperienceValue(character.level) +
+            // CharactersValueService.getAbilityValue(character.activeAbilityLevel) +
+            // CharactersValueService.getAbilityValue(character.passiveAbilityLevel) +
+            // CharactersValueService.getStarsValue(character.stars) -
+            // CharactersValueService.getInitialStarsValue(character.initialRarity) +
+            // CharactersValueService.getRarityValue(character.rarity) -
+            // CharactersValueService.getRarityValue(character.initialRarity) +
+            // not yet working:
+            CharactersValueService.getRankValue(character.name, character.rank, character.upgrades);
+        return Math.round(valueLevel);
+    }
+
+    public static getRankValue(name: string, currentRank: Rank, appliedUpgrades: string[]): number {
+        const CoinBS: number = 780 / 60000;
+        const MaterialBS: { [key: string]: number } = {
+            Common: 5,
+            Uncommon: 10,
+            Rare: 25,
+            Epic: 60,
+            Legendary: 150,
+        };
+
+        const upgrades = StaticDataService.getUpgrades({
+            id: name,
+            rankStart: Rank.Stone1,
+            rankEnd: currentRank,
+            appliedUpgrades: appliedUpgrades,
+        });
+
+        const [totalGold, setTotalGold] = useState<number>(0);
+
+        const allMaterials = useMemo<IMaterialRecipeIngredientFull[]>(() => {
+            const result: IMaterialRecipeIngredientFull[] = StaticDataService.groupBaseMaterials(upgrades, true);
+            const goldIndex = result.findIndex(x => x.id === 'Gold');
+
+            if (goldIndex > -1) {
+                const [gold] = result.splice(goldIndex, 1);
+                setTotalGold(gold.count);
+            } else {
+                setTotalGold(0);
+            }
+            return orderBy(result, ['rarity', 'count'], ['desc', 'desc']);
+        }, [upgrades]);
```

Comment:
The `useState` and `useMemo` should be used only inside of `tsx` files (components). Also `totalGold` is not used anywhere so it can be removed

```
 const result: IMaterialRecipeIngredientFull[] = StaticDataService.groupBaseMaterials(upgrades, true);
  const allMaterials = orderBy(result, ['rarity', 'count'], ['desc', 'desc']);
```